rust/datafusion/src/physical_plan/hash_join.rs

```diff
@@ -48,23 +54,15 @@ use super::{
 use crate::error::{DataFusionError, Result};
 
 use super::{ExecutionPlan, Partitioning, RecordBatchStream, SendableRecordBatchStream};
+use crate::physical_plan::coalesce_batches::concat_batches;
 use ahash::RandomState;
 use log::debug;
 
-// An index of (batch, row) uniquely identifying a row in a part.
-type Index = (usize, usize);
-// A pair (left index, right index)
-// Note that while this is currently equal to `Index`, the `JoinIndex` is semantically different
-// as a left join may issue None indices, in which case
-type JoinIndex = Option<(usize, usize)>;
-// An index of row uniquely identifying a row in a batch
-type RightIndex = Option<u32>;
-
 // Maps ["on" value] -> [list of indices with this key's value]
 // E.g. [1, 2] -> [(0, 3), (1, 6), (0, 8)] indicates that (column1, column2) = [1, 2] is true
 // for rows 3 and 8 from batch 0 and row 6 from batch 1.
-type JoinHashMap = HashMap<Vec<u8>, Vec<Index>, RandomState>;
-type JoinLeftData = Arc<(JoinHashMap, Vec<RecordBatch>)>;
```

Comment:
This change does mean that we can only use this join operator when the data can fit into a single batch. For example, if we are using 32-bit offsets then this limits us to 4GB of string data in a single column.

I don't know if we have any users that care about this scenario or not.

Reply:
Yes, but as far as I can tell only limits the max nr of elements to 2 ^ 32 for each batch on the right?
I think the string *data* per right record batch itself should be able to hold more than 4GB because the string offsets are stored separately, or am I wrong about that?
If this becomes a problem in the future, we can change it to use `u64` for the right indices as well, I think without too much extra cost besides double allocation/memory usage.

Reply:
The Arrow spec supports both 32-bit and 64-bit variants for the offset vector, so in the 32-bit case, each batch would be limited to 2^32 bytes of variable-width data per array.

Reply:
What I meant is that the types of the index used in the hash join has no effect of the type being used in the string array offsets. So the offsets values inside the string might be using 64-bit values, while here we use 32-bit to index into the string elements, which would allow more than 4GB? Anyway, at some point there is a limit we should keep in mind 👍

Reply:
Also FYI, for the left side we are now using 64 bit indices, so the 2^ 32 limit currently should be *only* on the number of elements per right batch, which seems reasonable to me (and easy to get rid of that constraint). Also, on the other hand, a future way to save some memory could be to choose for 32 bit indices for the left side whenever there are less than 2 ^ 32 items in the build side.

Reply:
Ah I see what you mean now...

Reply:
That would be a less trivial to get rid of than I thought, but at some point you could make a case either to convert those string batches to use 64 bit indexes or to split the batches to have 2^32 size, or something like that.

Reply:
Thanks for the explanation @andygrove 👍

Reply:
FWIW I would love to see a join approach for larger datasets that doesn't require giant buffers (eg. 64 bit indexes) but a runtime switch over to using sort/merge join or its equivalent. But that is a pipe dream for me at the moment.